File: server/csharp/MetaObjects.Render/Extract/Coercion.cs
namespace MetaObjects.Render.Extract;

/// <summary>
/// A recorded normalization/coercion event.
/// <paramref name="Kind"/> is a classification label, e.g. "alias", "clamp", "case", "runtime-alias-override".
/// <paramref name="From"/> and <paramref name="To"/> may be null when the transformation has no readable before/after value.
/// </summary>
public sealed record Coercion(string FieldPath, string? From, string? To, string Kind);

File: server/java/render/src/main/java/com/metaobjects/render/extract/Coercion.java
package com.metaobjects.render.extract;

import java.util.Objects;

/** A recorded normalization/coercion. kind e.g. "alias", "clamp", "case", "runtime-alias-override". */
public record Coercion(String fieldPath, String from, String to, String kind) {
    public Coercion {
        Objects.requireNonNull(fieldPath, "fieldPath");
        Objects.requireNonNull(kind, "kind");
    }
}
